### ISSUE TITLE
Additional structure for hybrid Post-Quantum procedures

### ISSUE THREAD

**Issue by RodriM11** [openssl/openssl#26279]
Hi! Regarding Post-Quantum Cryptography, it will be of paramount importance (at least in the subsequent years) the use of hybrid procedures, between "classical" (pre-quantum) and post-quantum procedures. As part of their use, it is of importance in certain situations (e.g. in hybrid KEM procedures, to later use some combiner, like a KDF or other, to derive the final hybrid shared secret) to be able to distinguish between the classical and PQ part of the cryptographic elements generated. In the case of public/private keys, since the structure holding them is an `EVP_PKEY`, it is already possible to do so (e.g., see `oqsprovider`). But, in the case of encapsulations, for instance, the API accounts directly for a `unsigned char *` result, for both the shared secret generated and the encapsulation value, and there is no way of separating both parts (without knowledge of the expected lengths, etc...). I think it would be useful to expand the API that PQ schemes will use to provide a more complete information of the hybrid elements generated. I imagine this could be done by either:

- Modifying the existing API to return more complex structures than a simple `unsigned char *`, in which information such as: lengths of each part, position of both classical and PQ, strength, etc... would be available
- Create a new API altogether for hybrid procedures, in which this new, more detailed structure would be employed.

Thanks in advance!

**Comment by t8m** [comment]
The hybrid combiner function would be part of the provided implementation. So I do not really see the need for this.

**Comment by RodriM11** [comment]
Thank you for the response. It is good to hear that it will be part of the new implementation. However, I still think there will be a lot of scenarios in which it will be desirable for users to be able to access each component, as they might be used/combined in a procedure that might be outside the scope of OpenSSL, or that it does not have as part of the provided implementation. Wouldn't that be a foreseeable case?

**Comment by baentsch** [comment]
> there will be a lot of scenarios in which it will be desirable for users to be able to access each component, as they might be used/combined in a procedure that might be outside the scope of OpenSSL

I think (precluding) this is precisely the reason for the answer by @t8m: Why should OpenSSL core logic cater for non-OpenSSL use cases? If someone wants to experiment with novel scenarios, use experimentation software (like `oqsprovider`); not opening up "heavy duty" software (like in `default provider`) for "other uses" should also in my opinion be avoided (or stated differently, it should stay focused on doing the thing it's been designed for -- hybrid TLS1.3 KEM ops in this case).

**Comment by RodriM11** [comment]
I agree with you that that should indeed be the target of OpenSSL, and maybe I didn't express myself properly. I meant that accessing those values (which are being generated under OpenSSL) separately could be of interest for applications that might be particular to their use (but using OpenSSL as the cryptographic engine for PQ hybrid procedures). If other software (such as `oqsprovider`) were valid choices, that should be a viable option, but since `oqsprovider` caters to OpenSSL's API, and this currently does not account for any way of accessing these value, I don't see a clean way to use it for this purpose. 

Nevertheless, if you don't consider it of relevance, I will close this issue. But, doesn't it seem somehow natural (while true that these hybrids will not be around forever) to be able to access these elements independently, in the new PQ setting?

**Comment by baentsch** [comment]
> could be of interest for applications that might be particular to their use

Could you point out such particular uses, @RodriM11 ? Of particular interest here would probably be reference to a standard that OpenSSL aims to implement.

> to be able to access these elements independently, in the new PQ setting

Again, what do you understand by "access"? Read-only properties for CLASSIC_PUBKEY/PRIVKEY? Or also "change-ability"? The OSSL_PARAM mechanism would allow that for `oqsprovider` even via unchanged `openssl` APIs. Maybe worth while taking the discussion to the [oqsprovider discussion board](https://github.com/open-quantum-safe/oqs-provider/discussions) unless someone here would like to keep it in this project (?)

**Comment by RodriM11** [comment]
I apologize, didn't see the answer. I was mainly thinking of uses (yes, mostly academic) of either new combiners that somehow somebody would like to implement and/or test (and that therefore will not be under the scope of OpenSSL) or for example combinations with other elements not pertaining  to OpenSSL (like QKD, etc...). But it is true that nonetheless, one could always generate each element independently (Classic, PQ, etc..) and treat each of the elements as its own, without the need for an additional structure.
By accessing I meant read-only properties. For example, once I have an encapsulation element generated from a hybrid procedure, be able to get the classical and PQ parts of it. Same with the shared secret generated. And I don't think this could be achieved now since the `encapsulation` API simply generates them as 'unsigned char *' buffers and not more complex structures. But maybe is too much trouble for the 'interest' it might generate.

**Comment by baentsch** [comment]
Thanks for the additional background @RodriM11 . In my eyes this means clearly things of more "research"/"novel use cases" which is the remit of `oqsprovider` and less of the `openssl` default/fips provider code base, right? Are you OK closing this here then and continue discussion in `oqsprovider` -- as I don't think any new `openssl` APIs are needed but the present OSSL_PARAMS mechanism can be utilized?

**Comment by RodriM11** [comment]
Absolutely. Thanks for your input and help!
